src/session_protocol.cpp

```diff
@@ -0,0 +1,651 @@
+            /*dest_sent_timestamp_ms=*/dest.sent_timestamp_ms,
+            /*dest_open_group_inbox_server_pubkey=*/dest.open_group_inbox_server_pubkey,
+            /*dest_closed_group_pubkey=*/dest.closed_group_pubkey,
+            /*dest_closed_group_keys=*/dest.closed_group_keys,
+            /*space=*/space,
+            /*use_malloc=*/UseMalloc::No);
+
+    EncryptedForDestination result = {
+            .encrypted = result_internal.encrypted,
+            .ciphertext = std::move(result_internal.ciphertext_cpp),
+    };
+    return result;
+}
+
+DecryptedEnvelope decrypt_envelope(
+        const DecryptEnvelopeKey& keys,
+        std::span<const uint8_t> envelope_payload,
+        std::chrono::sys_seconds unix_ts,
+        const array_uc32& pro_backend_pubkey) {
+    DecryptedEnvelope result = {};
+    SessionProtos::Envelope envelope = {};
+    std::span<const uint8_t> envelope_plaintext = envelope_payload;
+
+    // The caller is indicating that the envelope_payload is encrypted, if the group keys are
+    // provided. We will decrypt the payload to get the plaintext. In all other cases, the envelope
+    // is assumed to be unencrypted and can be used verbatim.
+    std::vector<uint8_t> envelope_plaintext_from_group_keys;
+    if (keys.use_group_keys) {
+        if (!keys.group_keys)
+            throw std::runtime_error(
+                    "API misuse: Envelope decryption with group keys was requested but no key was "
+                    "set");
+
+        // Decrypt
+        std::string sender_x25519_pubkey;
+        std::tie(sender_x25519_pubkey, envelope_plaintext_from_group_keys) =
+                keys.group_keys->decrypt_message(envelope_plaintext);
+        if (sender_x25519_pubkey.size() != crypto_sign_ed25519_PUBLICKEYBYTES)
+            throw std::runtime_error{fmt::format(
+                    "Parse encrypted envelope failed, extracted x25519 pubkey was wrong size: {}",
+                    sender_x25519_pubkey.size())};
+
+        // Update the plaintext to use the decrypted envelope
+        envelope_plaintext = envelope_plaintext_from_group_keys;  // Update the plaintext
+
+        // Copy keys out
+        std::memcpy(
+                result.sender_x25519_pubkey.data(),
+                sender_x25519_pubkey.data(),
+                sender_x25519_pubkey.size());
+    }
+
+    if (!envelope.ParseFromArray(envelope_plaintext.data(), envelope_plaintext.size()))
+        throw std::runtime_error{"Parse envelope from plaintext failed"};
+
+    // Parse type (unconditionallty)
+    if (!envelope.has_type())
+        throw std::runtime_error("Parse envelope failed, missing type");
+
+    switch (envelope.type()) {
+        case SessionProtos::Envelope_Type_SESSION_MESSAGE:
+            result.envelope.type = EnvelopeType::SessionMessage;
+            break;
+
+        case SessionProtos::Envelope_Type_CLOSED_GROUP_MESSAGE:
+            result.envelope.type = EnvelopeType::ClosedGroupMessage;
+            break;
+    }
```

Comment:
I think Android is the only one doing this now - just wondering whether we should properly deprecate this value and rely on the namespace that the message was received from (if a message with the wrong encryption type is stored in the wrong namespace then it should be considered invalid and dropped anyway)

Reply:
For sending we still need to as above for android but for decrypting on receive, you're saying remove this branch and replace it with:
```cpp
if (space == config::Namespace::GroupMessages) {
  result.envelop.type = EnvelopeType::GroupMessage;
} else {
  result.envelope.type = EnvelopeType::SessionMessage;
}
```
Is that right, where the C++ side still wants to keep the envelop type so application code after decryption has all necessary information self-contained in `struct Envelope` to work with, without requiring any external context. But requiring the `type` on the protobuf Envelope is no longer needed (and an outdated practice).

Reply:
Yea that way my thinking - since the decryption code is based on the `namespace` then we could eventually stop sending the `type` entirely once enough time has passed

> Is that right, where the C++ side still wants to keep the envelop type so application code after decryption has all necessary information self-contained in struct Envelope to work with, without requiring any external context.

I get where this is coming from but since the `Envelope` is encrypted for messages sent to a group the decryption logic is already dependent on knowing the `namespace` that a message came from which makes having a `Envelope.type` feel redundant 😕 (ie. at this point we already know that it's a `GroupMessage`)

Reply:
Alright. understood. I have removed the parsing of envelope types on decrypt.